domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -261,37 +346,387 @@ class TopicListController @Inject constructor(
     return recommendedStories
   }
 
-  private fun createRecommendedStoryFromAssets(topicId: String): PromotedStory? {
-    val topicJson = jsonAssetRetriever.loadJsonFromAsset("$topicId.json")!!
-    if (!topicJson.getBoolean("published")) {
-      // Do not recommend unpublished topics.
-      return null
+  private fun computePromotedActivityList(
+    topicProgressList: List<TopicProgress>
+  ): PromotedActivityList {
+    val promotedActivityListBuilder = PromotedActivityList.newBuilder()
+    if (topicProgressList.isNotEmpty()) {
+      promotedActivityListBuilder.promotedStoryList = computePromotedStoryList(topicProgressList)
+      if (promotedActivityListBuilder.promotedStoryList.getTotalPromotedStoryCount() == 0) {
+        promotedActivityListBuilder.comingSoonTopicList = computeComingSoonTopicList()
+      }
+    }
+    return promotedActivityListBuilder.build()
+  }
+
+  private fun computePromotedStoryList(topicProgressList: List<TopicProgress>): PromotedStoryList {
+    val recentlyPlayedStories = computePlayedStories(topicProgressList) { it < ONE_WEEK_IN_DAYS }
+    val olderPlayedStories = computePlayedStories(topicProgressList) { it > ONE_WEEK_IN_DAYS }
+    return PromotedStoryList.newBuilder()
+      .addAllRecentlyPlayedStory(
+        recentlyPlayedStories
+      )
+      .addAllOlderPlayedStory(
+        olderPlayedStories
+      )
+      .addAllSuggestedStory(
+        computeSuggestedStories(topicProgressList)
+      )
```

Comment:
```suggestion
    return PromotedStoryList.newBuilder()
      .addAllRecentlyPlayedStory(computePlayedStories(topicProgressList) { it < ONE_WEEK_IN_DAYS })
      .addAllOlderPlayedStory(computePlayedStories(topicProgressList) { it > ONE_WEEK_IN_DAYS })
      .addAllSuggestedStory(computeSuggestedStories(topicProgressList))
```

Can be collapsed slightly & still keep clarity, I think.